### PR TITLE
Allow use of independently-obtained access token (or full credentials)

### PR DESCRIPTION
**Pull request by rgrempel**
See #14

### CODE REVIEW

**Comment by rgrempel** [comment]
Essentially, this adds two new things (deliberately structured so that this is a "minor" change for semver purposes).

One new thing is:

```elm
{-| An alternative to `checkCachedCredentials` for cases where you don't have
the cached credentials, but you do have an access token, a `user`, or both.

  - If you supply the access token, we will contact the backend to verify that
    it is valid.
      - If that succeeds, we will be in `AuthenticatedUser` state, with
        credentials reflecting the supplied access token, and the `user` returned
        by the backend.
      - If that fails, then our state will depend on whether you supplied a
        `user` or not.
          - If you supplied a `user`, we will be in `AuthenticatedUser` state
            (with that `user`), but our `relogin` field will indicate the
            error checking the access token.
          - If you did not supply a `user`, we will be in `AnonymousUser`
            state, with the `progress` field indicating the error checking the
            access token.
  - If you don't supply an access token, then our state depends on whether you
    provide a `user` or not.
      - If you provide a `user`, we will be in `AuthenticatedUser` state with
        that user, but our `relogin` field will indicate that the access token
        was rejected.
      - If you didn't provide a `user`, we will be in `AnonymousUser` state.

Note that if you already have a `UserAndData` and you want to try a new access
token that you've obtained in one way or another, you should use
`tryAccessToken` instead.

-}
checkAccessToken : Config anonymousData user authenticatedData msg -> BackendUrl -> Maybe String -> Maybe user -> ( UserAndData anonymousData user authenticatedData, Cmd msg )
```

This is essentially an alternative to `checkCachedCredentials`, for cases where you don't have cached credentials, but you do have an access token or a user, in some other way.

The other new thing is `tryAccessToken`, which allows you to try a new access token in the context of an existing `UserAndData` object. So, you can either `tryLogin` to try a username and password, or `tryAccessToken` if you have some other way of getting an access token.

**Comment by rgrempel** [comment]
One thing I should make sure is documented well is that there are basically three kinds of things going on here.

1. Pure state

Both `loggedIn` and `loggedOut` are purely representations of state. They don't (by themselves) perform any action ... for instance, they don't check anything against the backend, and they don't cache credentials.

It's possible that we ought not expose `loggedIn` now. The idea behind exposing it was to deal with the use case described in #14 (i.e. cases where you obtain credentials in some other way). However, it doesn't really work well for that purpose, because we want to cache credentials (an action). So, we'd really want to use `checkAccessToken` in cases where you might be tempted to use `loggedIn`.

Exposing `loggedOut` probably still makes sense, being a sensible starting point in some cases. Though, if you are in fact logged in, one needs to use the `logout` message to actually logout of the backend, and destroy the cached credentials. So, I should make sure that is clearly documented.

2. Initialization

Both `checkCachedCredentials` and `checkAccessToken` are functions to be used to initialize a data structure. They don't depend on any existing `UserAndData` object -- you would use them to start from scratch.

Perhaps (when doing a breaking change) they ought to be renamed to reflect the idea of "initialization". In the meantime, I should make sure this is well documented.

3. Messages

Then we have `tryLogin`, `tryAccessToken` and `logout`, which are messages. That is, they update the state of an existing `UserAndData`.

**Comment by rgrempel** [comment]
This isn't tested yet -- that is, I haven't even tried it -- so it's not ready for review. But you could take a look at the new parts of the API to see if it seems generally nice.

**Comment by rgrempel** [comment]
I've now also added this one:

```elm
{-| Record a successful login which you've performed independently.

This differs from `loggedIn` in that it will cache the credentials you supply.

It differs from `checkAccessToken` in that it won't contact the backend to
verify the credentials you supply.

-}
recordLogin : Credentials user -> Msg user
```

Compared to what you can do with `checkAccesToken`, the main advantage of this is that it avoids an extra call to the backend in cases where you have both the access token and the user information ... we don't actually check the access token ... we just trust you've supplied valid credentials.

This still needs testing -- I haven't actually tried running any of this code yet.

**Comment by rgrempel** [comment]
I've done a pass over the documentation, which should now be much improved (I hope).

Next, I should do some testing, or at least try some examples -- this isn't actually so easy to test, without a certain number of complications.

File: tests/Restful/LoginTest.elm [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+exampleUser =
+    { userId = "example"
+    , name = "Example User"
+    }
+
+
+exampleCredentials : Credentials User
+exampleCredentials =
+    { accessToken = "adkfja;lskdj28d7skd"
+    , backendUrl = "http:/gizra.com/"
```
Comment (amitaibu) [inline tests/Restful/LoginTest.elm]: I think better `https://example.com`

**Comment by amitaibu** [comment]
`recordLogin` works nicely, and PR looks good. 👍

File: tests/Restful/LoginTest.elm (hunk 1 above)
Comment (rgrempel) [reply tests/Restful/LoginTest.elm]: I just visited https://example.com, and it turns out that it is an official example domain!

**Comment by rgrempel** [comment]
OK, I'll go ahead and merge it then!

**Comment by rgrempel** [comment]
I have published this, in version 2.1.0.

**Pull request merged**